Blue fire arrows

kr3z opened the pull request:
Adds a setting that allows ice arrows to melt red ice. Enabling the setting makes the following changes to the rom through Patches.py:
- Adds AT_TYPE_OTHER to the en_arrow actor's collider to allow collision with red ice. This change only affects collision with bg_ice_shelter (red ice actor) as this is the only actor that has AC_TYPE_OTHER without also having AC_TYPE_PLAYER. Arrows already have AT_TYPE_PLAYER so adding AT_TYPE_OTHER only adds collision with red ice.
- Removes collision for ice arrows for the second collision cylinder on bg_ice_shelter. Collision with the first cylinder is used for melting the red ice. Bg_ice_shelter types 0, 1, & 4 (various sized red ice crystals) also have a second collision cylinder which provides a hard surface to repel attacks. These two cylinders overlap each other and are exactly the same size. If ice arrows can collide with both, then collision will sometimes occur with the first cylinder and sometimes with the second. Removing ice arrow collision with the second cylinder means collision will always occur with the first cylinder, and reliably melt the red ice.
- Increases the radius/height of the collision cylinder for bg_ice_shelter type 3 (red ice sheets covering doors). The default size of the collision cylinder is much smaller than the ice sheet, which is fine for blue fire. But if the ice arrow hits the edges or top of the ice sheet, collision won't occur. The increased dimensions extend collision to cover the full ice sheet.

The asm code hooks into func_8089107C in z_bg_ice_shelter.c. The game code checks that collision occurred with the first cylinder of the red ice actor and then checks that the actor causing collision was blue fire. The asm code hooks into the actor check part and just extends it to check that the actor causing collision was either blue fire or an ice arrow.

Lastly, this change modifies the "Blue_Fire" & "_is_magic_arrow(item)" logic helpers so that ice arrows can be logically required when the blue_fire_arrows setting is enabled.

Demo video of blue fire arrows in action:
https://youtu.be/5usbP-QHuNw

kr3z committed:
Blue Fire Arrows

Ice arrows melt red ice

kr3z committed:
Moved setting to misc category

Moved setting to misc category

kr3z committed:
Updated LogicHelpers.json

Fix to blue fire arrows in LogicHelpers.json

r0bd0g commented:
If you want this to be logical you also have to change when ice arrows are able to be foolish. But I feel like this ought not to be logical and it's just a thing you can enable on the memes tab. Time goes on and candidates for the memes tab continue to grow...

matthewkirby commented:
I love this idea! =D

Per chance will these work on the mud walls in Dodongo's Cavern?

kr3z committed:
Reverted logic changes

Reverted changes to LogicHelpers.json so that blue fire arrows are never in logic. Also removed the randomize_key param from the setting in SettingsList.py. This should have been removed in the last commit when I moved the setting to the Other tab.

kr3z commented:
I removed the logic for blue fire arrows. I had forgotten that blue fire can break mud walls. I see that is considered a logical use of blue fire in glitched logic, which means the changes I had made to the Blue_Fire logic helper would break glitched logic. So let's just go with blue fire arrows being out of logic.

As for the question about the mud walls, my intention was to only allow ice arrows to melt red ice. The current code does not affect the bg_breakwall actor. Though, I wouldn't mind getting more feedback on this. If the community consensus is that blue fire arrows should be able to break the mud walls (since blue fire can), then I can do that. That change should be pretty simple.

LostThePirate commented:
If you're making Blue Fire Arrows, I feel like you may as well go all in and give them the same utility as bottled Blue Fire. I'm in the camp that says give them the ability to break mud walls.

kr3z committed:
Blue fire arrows break mud walls

Enabling setting sets ice arrow bumper flag on bg_breakwall so that ice arrows can be used to break mud walls. Also updated tooltip on setting to note this.

kr3z commented:
I added the ability to break mud walls. The change adds one more line to Patches.py which sets the ice arrow bit in the bumper flags on the bg_breakwall actor. Also updated the tooltip description for the setting to note that blue fire arrows can both melt red ice and break mud walls.

r0bd0g commented:
I think the bombable walls in dmt are also susceptible? And the bombable floor above the boss in DC. (I dunno if maybe those are all the same actor or not.)

kr3z commented:
The bombable walls in DMT are bg_bombwall actors which do not have AC_TYPE_OTHER and so don't have collision with blue fire (I also just tested this in gz to be sure, and blue fire does not affect the dmt walls).

The floor above king dodongo is a bg_breakwall actor and the change I made does enable collision with that. The collision is a bit annoying to hit with ice arrows, though. I'd say its similar to doing it with hammer. You basically have to stand on top of it and shoot down, and it may take a couple of tries to find the right spot. 

I took a look through all the actors that have collision with blue fire to double check that I've covered everything. The following actors have AC_TYPE_OTHER collision:
- bg_breakwall: has been updated to have collision with ice arrows
- bg_ice_shelter: has been updated to have collision with ice arrows
- en_bom: actor for bombs. Already has collision with ice arrows so no update needed
- en_bombf: actor for bombflowers. Already has collision with ice arrows so no update needed

And the following actors have AC_TYPE_ALL collision, which includes AC_TYPE_OTHER:
- en_a_obj: It doesn't look like blue fire would affect these objects, so no change should be needed here
- bg_ddan_kd: The stone stairs in DC. Not affected by blue fire, so no change should be needed
- bg_dodoago: The big dodongo head in DC. Again, not affected by blue fire so no change should be needed

So the changes I've made should give ice arrows all the same abilities as blue fire.

cjohnson57 commented:
Personally I'm not really a fan of the idea of a meme tab... I'd like all settings added to have logic support

r0bd0g commented:
OHKO was added just on a whim with no thought about the logical implications. It has full logic support now but it's been one of the more annoying settings to keep the logic maintained for. I think that's cue enough to let memes be memes.

fenhl commented:
Voicing my support for adding this into logic. Now that the ability to break mud walls has been added, there are no known issues with adding it, and since this setting essentially just makes one item act like another, it's unlikely to cause issues.

fenhl commented on SettingsList.py:
```diff
@@ -5250,6 +5250,17 @@ def __init__(self, name, gui_text, min, max, default, step=1,
+        gui_tooltip    = '''\
+            Ice arrows gain the power of blue fire.
+            They can be used to melt red ice
+            and break the mud walls in Dodongos Cavern
```
```suggestion
            and break the mud walls in Dodongo's Cavern.
```

GSKirox commented:
I think there should be an easy in game indicator that instantly tells if the setting is on, other than just throwing an arrow at red ice or mud walls.
The easiest way imo would be to change the name of the item from Ice Arrows to Blue Fire Arrows, that way you'll know either when you get them (can look at the inventory if you mashed the text too fast or if you're in text shuffled), or if you get a hint for it.

August40th commented:
I think Kirox suggestion to change the name is a good one, another way to know the setting status would to change the color. Hard decision: should it be the purple color of a bottled flame or ice blue color of the torches & emptied flames ?

I'd vote the latter

kr3z committed:
Text updates

Updated tooltip
Changed item text when receiving item and in inventory

mracsys commented:
If the only concern with having these in logic is breaking glitched logic, the setting could be added to the long list of glitched-incompatible settings. Glitchless logic is easy to change, and I'd love a meme seed with ice arrows as way of the hero.

cjohnson57 commented:
After a conversation on the Discord, the majority opinion seems to be okay with having this in the randomizer.

Updating the shop/hint text would be good, but mainly, once this is added back to logic I'll consider this PR good to go.

r0bd0g commented:
I really don't think we need this setting. But after the shield in the pot I think the overton window has moved.

cjohnson57 commented:
@kr3z Are you planning to update the PR to add the BFAs back to logic?

kr3z committed:
Re-add logic

Re-added logic for blue fire arrows

kr3z commented:
Logic re-added

cjohnson57 commented:
Thanks! I'll review it sometime.
